# Network analyses

Here I employed two methods that will output a species network, since we suspected the occurrence of gene flow between the species in our dataset. We used two approaches: PhyloNet and Treemix.

## PhyloNet (InferNetwork_MPL)

Due to the size of our dataset (15 species, 8889 gene trees), I used the [pseudo-maximum likelihood approach](https://wiki.rice.edu/confluence/display/PHYLONET/InferNetwork_MPL) available in PhyloNet to model a species network with reticulation events. I ran several tests with different subsets of the gene trees and found that the best results were obtained when I used all gene trees available. I could not succesfully run the full maximum likelihood approach with my dataset, but there are ways of doing it where PhyloNet can be ran several times with different subsets of the gene trees (e.g., [Edelman et al 2019](https://science.sciencemag.org/content/366/6465/594.full)). 

Using the **rooted** gene trees generated during the [ASTRAL analysis](https://github.com/MafaldaSFerreira/hare-phylogenomics/blob/master/3.species_tree_analysis/3.species_tree_analysis.md#astral), use the R package `ape` to conver them to nexus format.

```
library(ape)
trees<-read.tree("gene_trees_TCs5.tree")
write.nexus(trees,file="gene_trees_TCs5.nex",translate=F)
```

This bit of code only takes us halfway to the PhyloNet input. Use `convert_gene_trees_to_nex.py` to format the gene trees as PhyloNet wants them. This command will also add the final block of code that is the actual PhyloNet command. After the input and output files names, add the PhyloNet command you which to run.

```
convert_gene_trees_to_nex.py gene_trees_TCs5.nex gene_trees_TCs5_m0_plopt.nex "InferNetwork_MPL (all) 0 -o -pl 10 -a <B_idahoensis:BID_RC_444; L_americanus:LAM_A0604_WA,LAM_A0965_WA,LAM_Allab3_Bo,LAM_NBerg67_Rockies; L_californicus:LCF_CAL_1954,LCF_TUC_2060; L_callotis:LCL_1956; L_capensis:LCP_SAF_1903,LCP_TAN_3023; L_corsicanus:LCR_ITA_1957,LCR_ITA_1958; L_castroviejoi:LCS_1891,LCS_1894; L_europaeus:LER_PYR_1546,LER_VIE_1639; L_fagani:LFG_FAG114,LFG_FAG96; L_granatensis:LGR_CRE_2553,LGR_SEV_1163; L_habessinicus:LHB_HAB35,LHB_HAB68; L_mandschuricus:LMS_PRI_2460,LMS_PRI_2461; L_starki:LST_STA65,LST_STA89; L_othus:LOT_2109,LOT_2110; L_timidus:LTM_AFR_3108,LTM_CAT_2012,LTM_MAG_1862; L_townsendii:LTW_JRRK_3,LTW_MTA_3280; O_cuniculus:OryCun>"
convert_gene_trees_to_nex.py gene_trees_TCs5.nex gene_trees_TCs5_m1_plopt.nex "InferNetwork_MPL (all) 1 -o -pl 10 -a <B_idahoensis:BID_RC_444; L_americanus:LAM_A0604_WA,LAM_A0965_WA,LAM_Allab3_Bo,LAM_NBerg67_Rockies; L_californicus:LCF_CAL_1954,LCF_TUC_2060; L_callotis:LCL_1956; L_capensis:LCP_SAF_1903,LCP_TAN_3023; L_corsicanus:LCR_ITA_1957,LCR_ITA_1958; L_castroviejoi:LCS_1891,LCS_1894; L_europaeus:LER_PYR_1546,LER_VIE_1639; L_fagani:LFG_FAG114,LFG_FAG96; L_granatensis:LGR_CRE_2553,LGR_SEV_1163; L_habessinicus:LHB_HAB35,LHB_HAB68; L_mandschuricus:LMS_PRI_2460,LMS_PRI_2461; L_starki:LST_STA65,LST_STA89; L_othus:LOT_2109,LOT_2110; L_timidus:LTM_AFR_3108,LTM_CAT_2012,LTM_MAG_1862; L_townsendii:LTW_JRRK_3,LTW_MTA_3280; O_cuniculus:OryCun>"
convert_gene_trees_to_nex.py gene_trees_TCs5.nex gene_trees_TCs5_m2_plopt.nex "InferNetwork_MPL (all) 2 -o -pl 10 -a <B_idahoensis:BID_RC_444; L_americanus:LAM_A0604_WA,LAM_A0965_WA,LAM_Allab3_Bo,LAM_NBerg67_Rockies; L_californicus:LCF_CAL_1954,LCF_TUC_2060; L_callotis:LCL_1956; L_capensis:LCP_SAF_1903,LCP_TAN_3023; L_corsicanus:LCR_ITA_1957,LCR_ITA_1958; L_castroviejoi:LCS_1891,LCS_1894; L_europaeus:LER_PYR_1546,LER_VIE_1639; L_fagani:LFG_FAG114,LFG_FAG96; L_granatensis:LGR_CRE_2553,LGR_SEV_1163; L_habessinicus:LHB_HAB35,LHB_HAB68; L_mandschuricus:LMS_PRI_2460,LMS_PRI_2461; L_starki:LST_STA65,LST_STA89; L_othus:LOT_2109,LOT_2110; L_timidus:LTM_AFR_3108,LTM_CAT_2012,LTM_MAG_1862; L_townsendii:LTW_JRRK_3,LTW_MTA_3280; O_cuniculus:OryCun>"
convert_gene_trees_to_nex.py gene_trees_TCs5.nex gene_trees_TCs5_m3_plopt.nex "InferNetwork_MPL (all) 3 -o -pl 10 -a <B_idahoensis:BID_RC_444; L_americanus:LAM_A0604_WA,LAM_A0965_WA,LAM_Allab3_Bo,LAM_NBerg67_Rockies; L_californicus:LCF_CAL_1954,LCF_TUC_2060; L_callotis:LCL_1956; L_capensis:LCP_SAF_1903,LCP_TAN_3023; L_corsicanus:LCR_ITA_1957,LCR_ITA_1958; L_castroviejoi:LCS_1891,LCS_1894; L_europaeus:LER_PYR_1546,LER_VIE_1639; L_fagani:LFG_FAG114,LFG_FAG96; L_granatensis:LGR_CRE_2553,LGR_SEV_1163; L_habessinicus:LHB_HAB35,LHB_HAB68; L_mandschuricus:LMS_PRI_2460,LMS_PRI_2461; L_starki:LST_STA65,LST_STA89; L_othus:LOT_2109,LOT_2110; L_timidus:LTM_AFR_3108,LTM_CAT_2012,LTM_MAG_1862; L_townsendii:LTW_JRRK_3,LTW_MTA_3280; O_cuniculus:OryCun>"
convert_gene_trees_to_nex.py gene_trees_TCs5.nex gene_trees_TCs5_m4_plopt.nex "InferNetwork_MPL (all) 4 -o -pl 10 -a <B_idahoensis:BID_RC_444; L_americanus:LAM_A0604_WA,LAM_A0965_WA,LAM_Allab3_Bo,LAM_NBerg67_Rockies; L_californicus:LCF_CAL_1954,LCF_TUC_2060; L_callotis:LCL_1956; L_capensis:LCP_SAF_1903,LCP_TAN_3023; L_corsicanus:LCR_ITA_1957,LCR_ITA_1958; L_castroviejoi:LCS_1891,LCS_1894; L_europaeus:LER_PYR_1546,LER_VIE_1639; L_fagani:LFG_FAG114,LFG_FAG96; L_granatensis:LGR_CRE_2553,LGR_SEV_1163; L_habessinicus:LHB_HAB35,LHB_HAB68; L_mandschuricus:LMS_PRI_2460,LMS_PRI_2461; L_starki:LST_STA65,LST_STA89; L_othus:LOT_2109,LOT_2110; L_timidus:LTM_AFR_3108,LTM_CAT_2012,LTM_MAG_1862; L_townsendii:LTW_JRRK_3,LTW_MTA_3280; O_cuniculus:OryCun>"
convert_gene_trees_to_nex.py gene_trees_TCs5.nex gene_trees_TCs5_m5_plopt.nex "InferNetwork_MPL (all) 5 -o -pl 10 -a <B_idahoensis:BID_RC_444; L_americanus:LAM_A0604_WA,LAM_A0965_WA,LAM_Allab3_Bo,LAM_NBerg67_Rockies; L_californicus:LCF_CAL_1954,LCF_TUC_2060; L_callotis:LCL_1956; L_capensis:LCP_SAF_1903,LCP_TAN_3023; L_corsicanus:LCR_ITA_1957,LCR_ITA_1958; L_castroviejoi:LCS_1891,LCS_1894; L_europaeus:LER_PYR_1546,LER_VIE_1639; L_fagani:LFG_FAG114,LFG_FAG96; L_granatensis:LGR_CRE_2553,LGR_SEV_1163; L_habessinicus:LHB_HAB35,LHB_HAB68; L_mandschuricus:LMS_PRI_2460,LMS_PRI_2461; L_starki:LST_STA65,LST_STA89; L_othus:LOT_2109,LOT_2110; L_timidus:LTM_AFR_3108,LTM_CAT_2012,LTM_MAG_1862; L_townsendii:LTW_JRRK_3,LTW_MTA_3280; O_cuniculus:OryCun>"

```

Finally, run PhyloNet:

```
java -jar ~/programs/PhyloNet_3.6.6.jar gene_trees_TCs5_m4_plopt.nex > gene_trees_TCs5_m4_plopt.out
```

For each model we calculated AIC, AICc and BIC following the recommendations of [Yu et al 2012](https://doi.org/10.1371/journal.pgen.1002660) and [Yu et al 2014; see Appendix for formulas](https://doi.org/10.1073/pnas.1407950111).

## TreeMix

[TreeMix](https://bitbucket.org/nygcresearch/treemix/wiki/Home) is a method for inferring the patterns of population splits and mixtures in the history of a set of populations. It requires allele counts for each population (in our case, species) as input.

The pipeline here will be similar to the one used for [SVDquartets](3.species_tree_analysis/create_svdquartets_input.sh). The exceptions are that we use _L. callotis_ as our outgroup. We also do not allow missing data in our set of SNPs so `filter_missing_vcf.py` is set to zero.

The script we will use here will be `create_input_treemix.sh`. It will take the consensus chromosome fastas for each individual and concatenate them into a multi individual fasta alignment which is used as input for `snp-sites`. This program will call snps in your alignment and generate a vcf file. Using `bedtools`, we keep only the SNPs inside target regions extended by 200bp. Then, we choose a SNP every 10kb. In the end, we will have a single fasta alignment for each chromosome, where each entry is a sequence of concatened SNPs for each sample. The script requires:

1. Whole chromosome consensus fasta sequences for each individual generated [previously](https://github.com/MafaldaSFerreira/hare-phylogenomics/blob/master/3.species_tree_analysis.md#1-generate-consensus-fasta-per-individual-and-chromosome-as-described-here). Alter the paths to access them and maintain a `samples.txt` file with sample IDs in your folder.
2. `snp-sites`; this script with convert a whole chromosome fasta alignment to vcf.
3. `filter_missing_vcf.py`; filters SNPs from vcf that have too much missing data (adjustable)
4. `filter_bed.py`; limits the final alignment to 1 SNP per 10kb.
5. `bedtools`
6. `concatenate_snps.py`; concatenates the output of `bedtools getfasta`

Run it like so, with a file containing the list of `target_chromosomes.txt`(all chromosomes contained capture targets).

```
while read -r LINE; do create_input_treemix.sh "$LINE"; done < target_chromosomes.txt
```

Some of the generated fasta alignments will be empty (no SNPs were called). Here is some code to find and delete these empty alignments.

```
# find empty files:
ls -lhaS
for i in $(ls *.fa); do samtools faidx $i; done
for i in $(find . -maxdepth 1 -name "*.fa" -size 449c -type f -exec ls {} \;); do head -n1 "$i".fai ; done
find . -maxdepth 1 -name "*.fa" -size 449c -type f -delete
```

Run `AMAS` to concatenate all chromosome alignments into a single fasta alignment.

```
AMAS.py concat -i *.fa -f fasta -d dna -t snp_alignment_LepusOnly.fas -p snp_alignment_LepusOnly.txt -u fasta -y unspecified
```

Ok, so... `TreeMix` requires allele frequencies as input and I used this wonderful script [`treemix_from_nex.py`](https://github.com/mgharvey/misc_python/blob/master/bin/TreeMix/treemix_from_nex.py) from [Michael Harvey](https://github.com/mgharvey) that converts a fasta file into `TreeMix` input. However, it requires phased positions. Therefore, we are going to randomly phase positions using `fake_phase_fasta_algn.py`. For each sample, this script will create two fasta entries "_a" and "_b" with randomly phased heterozygout sites.

```
fake_phase_fasta_algn.py snp_alignment_LepusOnly.fas snp_alignment_LepusOnly_hap.fas
```

Then, create the `TreeMix` input. The first number is the number of populations (15 species), and then following numbers are the number of individuals in each population. We use the number of chromosomes. `treemix_from_nex.py` will remove tri-allelic sites, which is very convenient :).

```
treemix_from_nex.py snp_alignment_LepusOnly_hap.fas snp_alignment_LepusOnly_hap_treemix_input 15 8 4 2 4 4 4 4 4 4 4 4 4 4 6 4
```

You can change the header of the output to contain the names of your populations. Otherwise, they are named "pop1", "pop2", "pop3", etc... Gzip the file (`TreeMix` requires gzip inputs). And then run `TreeMix`. These were the commands I used:

```
~/my_programs/treemix-1.13/src/treemix -i snp_alignment_LepusOnly_hap_treemix_inputtreemix_file_out.txt.gz -global -noss -o treemix_LepusOnly_m0 -root LCL -se
~/my_programs/treemix-1.13/src/treemix -i snp_alignment_LepusOnly_hap_treemix_inputtreemix_file_out.txt.gz -global -noss -o treemix_LepusOnly_m1 -root LCL -m 1 -se
~/my_programs/treemix-1.13/src/treemix -i snp_alignment_LepusOnly_hap_treemix_inputtreemix_file_out.txt.gz -global -noss -o treemix_LepusOnly_m4 -root LCL -m 4 -se
~/my_programs/treemix-1.13/src/treemix -i snp_alignment_LepusOnly_hap_treemix_inputtreemix_file_out.txt.gz -global -noss -o treemix_LepusOnly_m5 -root LCL -m 5 -se
~/my_programs/treemix-1.13/src/treemix -i snp_alignment_LepusOnly_hap_treemix_inputtreemix_file_out.txt.gz -global -noss -o treemix_LepusOnly_m6 -root LCL -m 6 -se
~/my_programs/treemix-1.13/src/treemix -i snp_alignment_LepusOnly_hap_treemix_inputtreemix_file_out.txt.gz -global -noss -o treemix_LepusOnly_m7 -root LCL -m 7 -se
~/my_programs/treemix-1.13/src/treemix -i snp_alignment_LepusOnly_hap_treemix_inputtreemix_file_out.txt.gz -global -noss -o treemix_LepusOnly_m8 -root LCL -m 8 -se
~/my_programs/treemix-1.13/src/treemix -i snp_alignment_LepusOnly_hap_treemix_inputtreemix_file_out.txt.gz -global -noss -o treemix_LepusOnly_m9 -root LCL -m 9 -se
```
